src/hotspot/share/utilities/compilerWarnings_gcc.hpp

```diff
@@ -70,32 +70,19 @@
 #define PRAGMA_ZERO_AS_NULL_POINTER_CONSTANT_IGNORED \
   PRAGMA_DISABLE_GCC_WARNING("-Wzero-as-null-pointer-constant")
 
-#if (__GNUC__ >= 10)
-// TODO: Re-enable warning attribute for Clang once
-// https://github.com/llvm/llvm-project/issues/56519 is fixed and released.
-// || (defined(__clang_major__) && (__clang_major__ >= 14))
-
-// Use "warning" attribute to detect uses of "forbidden" functions.
-//
-// Note: The warning attribute is available since GCC 9, but disabling pragmas
-// does not work reliably in ALLOW_C_FUNCTION. GCC 10+ and up work fine.
-//
-// Note: _FORTIFY_SOURCE transforms calls to certain functions into calls to
-// associated "checking" functions, and that transformation seems to occur
-// *before* the attribute check.  We use fortification in fastdebug builds,
-// so uses of functions that are both forbidden and fortified won't cause
-// forbidden warnings in such builds.
-#define FORBID_C_FUNCTION(signature, alternative) \
-  extern "C" __attribute__((__warning__(alternative))) signature;
-
-// Disable warning attribute over the scope of the affected statement.
-// The name serves only to document the intended function.
-#define ALLOW_C_FUNCTION(name, ...)                     \
-  PRAGMA_DIAG_PUSH                                      \
-  PRAGMA_DISABLE_GCC_WARNING("-Wattribute-warning")     \
-  __VA_ARGS__                                           \
-  PRAGMA_DIAG_POP
-
-#endif // gcc10+
+#define PRAGMA_DEPRECATED_IGNORED \
+  PRAGMA_DISABLE_GCC_WARNING("-Wdeprecated-declarations")
+
+// This variant of FORBID_C_FUNCTION overrides the default variant.  Clang
+// seems to make a distinction between [[noreturn]] and the old-style noreturn
+// attribute.  For example, if <stdlib.h> has already been included, using
+// [[noreturn]] when forbidding exit(int) gives "error: 'noreturn' attribute
+// does not appear on the first declaration", with the previous declaration
+// reported as being "void exit(int) __dead2;". __dead2 is an old-style
+// noreturn attribute.
+#ifdef __clang__
+#define FORBID_NORETURN_C_FUNCTION(Signature, Alternative) \
+  FORBID_C_FUNCTION(__attribute__((__noreturn__)) Signature, Alternative)
```

Comment:
clang isn't the only one that makes a distinction between [[gnu::noreturn]] and [[noreturn]], gcc does too. It's a little strange that gcc doesn't face this same issue

Reply:
I don't know what `[[gnu::noreturn]]` has to do with this?  And in what way does gcc treat them differently?

Reply:
[[gnu::noreturn]] is equal to \_\_attribute\_\_((noreturn)), I just referred to it as its scoped attribute form to make it clear that I wasn't talking about the Standard noreturn. There is funnily enough no documentation on how gcc treats them differently, I only learnt that it does while reading through gcc's source code one day

```cpp
/* We used to treat C++11 noreturn attribute as equivalent to GNU's,
	 but no longer: we have to be able to tell [[noreturn]] and
	 __attribute__((noreturn)) apart.
	 Similarly for C++14 deprecated attribute, we need to emit extra
	 diagnostics for [[deprecated]] compared to [[gnu::deprecated]].  */
      /* C++17 fallthrough attribute is equivalent to GNU's.  */
```

Reply:
gcc does treat [[noreturn]] differently from [[gnu::noreturn]] /
attribute((noreturn)) when there is a preceding declaration that doesn't have
either kind of attribute.  Compile this:
```
void frob(int);
//__attribute__((__noreturn__)) void frob(int);
//[[gnu::noreturn]] void frob(int);
[[noreturn]] void frob(int);
```
and you'll get an error:
```
error: function 'void frob(int)' declared '[[noreturn]]' but its first declaration was not
```
But uncomment either of the gnu-specific attributes and the error goes away.

C++14 7.6.3 "Noreturn attribute" says "The first declaration of a function
shall specify the noreturn attribute if any declaration of that function
specifies the noreturn attribute."

So without the declarations with gcc-specific attributes it errors, and that's
fine. The gcc-specific attributes never complained in that situation, and
continue to not complain, for backward compatibility. But other than that
error checking difference, the gcc-specific attributes seem to be treated as
equivalent to the standard attribute.

Reply:
The difference between [[deprecated]] and [[gnu::deprecated]] is discussed here:
https://gcc.gnu.org/git/?p=gcc.git;a=commit;f=gcc/cp/parser.cc;h=cd647514a539943ade6461efbf056a7c3f4305c6
TL;DR - [[gnu::deprecated]] is allowed in more places than [[deprecated]].